Issue 57 bis - Closes #64

gggeek opened the pull request:
A fix for issue #57 that avoids introducing issue #64.
Intended to replace https://github.com/instaclick/php-webdriver/pull/58

aik099 commented:
1. you can really squash at this point
2. please add `Closes #64` to PR description so that associated issue will be closed automatically on PR merge

aik099 commented on composer.json:
```diff
@@ -27,7 +27,8 @@
     "require-dev": {
-        "satooshi/php-coveralls": "dev-master"
+        "satooshi/php-coveralls": "dev-master",
+        "phpunit/phpunit": ">=4.0.0"
```
How this is helpful knowing, that:

* Travis CI has PHPUnit installed already
* Most of developers have PHPUnit installed anyway (e.g. via Composer globally or PHAR)

gggeek replied:
well, I personally do not have phpunit installed globally, but tend to install it inside every project I develop. This because there have been in the past incompatible updates, so different projects were stuck with different phpunit versions.
But I can move this to a separate PR if you think

aik099 replied:
> well, I personally do not have phpunit installed globally, but tend to install it inside every project I develop.

I'm personally against having tools, that are better to be used at latest version on all projects as project dependency at all. PHPUnit tends to include more and more different dependencies to work and all that really takes up disk space.

gggeek replied:
ok then, will revert

aik099 commented on lib/WebDriver/AbstractWebDriver.php:
```diff
@@ -122,15 +124,30 @@ protected function curl($requestMethod, $command, $parameters = null, $extraOpti
 
         list($rawResult, $info) = ServiceFactory::getInstance()->getService('service.curl')->execute($requestMethod, $url, $parameters, $extraOptions);
 
+        // According to the spec:
```
I'm sure this huge comment can be shortened like: `According to <link here> all 4xx codes are considered as an error`.

gggeek replied:
can shorten it.
since this weird behaviour is where it all started, I thought a detailed comment was appropriate

gggeek replied:
done

aik099 commented on lib/WebDriver/AbstractWebDriver.php:
```diff
@@ -122,15 +124,30 @@ protected function curl($requestMethod, $command, $parameters = null, $extraOpti
+        // - invalid requests (http 4xx) are instead plaintext
+        // See https://w3c.github.io/webdriver/webdriver-spec.html for details
+
+        if ($info['http_code'] >= 400 && $info['http_code'] <= 499) {
```
1. Better to do `< 500` rather then `<= 499` to clearly indicate what we need.
2. Maybe we can just check for 1st digit in number to be `4`?

gggeek replied:
- can do < 500
- I did not rely on 1st digit only because I was unsure about status codes > 999 being valid, but at least according to rfc7231 they are not, so can do

gggeek replied:
done

aik099 commented on lib/WebDriver/AbstractWebDriver.php:
```diff
@@ -122,15 +124,30 @@ protected function curl($requestMethod, $command, $parameters = null, $extraOpti
+        // - valid responses always include: sessionId, status and value
+        //   (at least as long as 'they include a message body' - but we tested commands such as timeouts/async_script
+        //   and session/delete and they still give back a json payload)
+        // - failed commands (http 500) include those as well
```
Why we mention error that we don't handle, like 500?

gggeek replied:
well, we do handle, implicitly, by swallowing it

aik099 commented on lib/WebDriver/AbstractWebDriver.php:
```diff
@@ -122,15 +124,30 @@ protected function curl($requestMethod, $command, $parameters = null, $extraOpti
+        // See https://w3c.github.io/webdriver/webdriver-spec.html for details
+
+        if ($info['http_code'] >= 400 && $info['http_code'] <= 499) {
+            throw WebDriverException::factory(Exception::CURL_EXEC, 'Webdriver http error: ' . $info['http_code'] . ', payload :' . substr($rawResult, 0, 1000)/*, $previousException*/);
```
Is `/*, $previousException*/` really needed?

gggeek replied:
1000 is indeed completely arbitrary. And the unattentive user might be caught off guard (I should have really introduced at least shortening with ... if resp. length was over 1000).
It's just that http responses can be massive, and I thought having no limit was risky

aik099 commented on lib/WebDriver/AbstractWebDriver.php:
```diff
@@ -122,15 +124,30 @@ protected function curl($requestMethod, $command, $parameters = null, $extraOpti
-        $value   = null;
 
-        if (is_array($result) && array_key_exists('value', $result)) {
+        if (!is_array($result) || !array_key_exists('status', $result)) {
```
If `$result` is `NULL`, then it means that `json_decode` failed to decode the string and using `json_last_error()` (PHP 5.3+) will get us actual decoding error (if we ever needed it).

In which cases `$rawResult` was valid JSON, but `$result` isn't an array with `status` key?

gggeek replied:
Json without status is not sent back from Selenium, at least in my cursory testing.
But I wondered about the rare case where an intermediate proxy would also send back a valid json, or otherwise mess with the response, and thought that extra safety is good.

In fact in the curl info we have the content-type as well, so I might:
- avoid json-decode altogether if wrong content-type is found, and give a specific error msg
- use json_last_error in case json decoding fails to have the best possible error message

aik099 replied:
> avoid json-decode altogether if wrong content-type is found, and give a specific error msg

Better not to rely on `Content-Type` header for security reasons. For now we presume, that on the other end we'll be having valid Selenium server and lets hope this would always be the case.

gggeek replied:
changed: separate error messages for cases of non-json and non-array or missing-status

aik099 commented on lib/WebDriver/AbstractWebDriver.php:
```diff
@@ -122,15 +124,30 @@ protected function curl($requestMethod, $command, $parameters = null, $extraOpti
 
-        if (is_array($result) && array_key_exists('value', $result)) {
+        if (!is_array($result) || !array_key_exists('status', $result)) {
+            throw WebDriverException::factory(Exception::CURL_EXEC, 'Payload received from webdriver is not valid json: ' . substr($rawResult, 0, 1000)/*, $previousException*/);
```
Same notes as for previous exception throwing place.

aik099 commented on lib/WebDriver/Service/CurlService.php:
```diff
@@ -87,21 +88,28 @@ public function execute($requestMethod, $url, $parameters = null, $extraOptions
             );
-
-            throw WebDriverException::factory(WebDriverException::CURL_EXEC, $message . "\n\n" . $error);
+            $detailsException = new CurlDetailsException($error, $errno, null, $rawResult, $info);
```
Why don't we throw this exception? The `CurlService` seems to be standalone component and isn't required to use `WebDriverException::factory` method.

gggeek replied:
Agree.
I just tried to keep maximum BC

aik099 commented on test/CI/Travis/apache_vhost:
```diff
@@ -0,0 +1,68 @@
+# Configuration file for Apache running on Travis.
```
Are you sure, that such complex setup is needed. Here is how the Selenium tests are setup at Mink: https://github.com/minkphp/MinkSelenium2Driver/blob/master/.travis.yml#L27-L30

gggeek replied:
Not really, I just took a vhost I had used on Travis for another project (you can see I even forgot to remove the php part).

The custom vhost gives us more room to expand in the future, but I can go for the shorter version if preferred

gggeek replied:
vhost config removed

aik099 commented on test/CI/Travis/setup_selenium.sh:
```diff
@@ -0,0 +1,7 @@
+
+# set up Selenium for functional tests
+
+wget http://goo.gl/cvntq5 -O selenium.jar
```
Why not use direct url to seleniumhq website for download?

gggeek replied:
this is the question I mentioned in the main issue: which version of selenium to test against? always the latest one, or a fixed one? or maybe one different selenium version for each php version?

spolischook replied:
I vote to test with only one version and add it to composer.json

aik099 commented on test/Test/WebDriver/WebDriverTest.php:
```diff
@@ -53,16 +61,37 @@ protected function tearDown()
             $this->session = $this->driver->session();
         } catch (\Exception $e) {
-            if (strpos($e->getMessage(),'Failed connect to localhost:4444; Connection refused') !== false
+            if (preg_match('/Failed to connect to .* Connection refused/', $e->getMessage()) != false
```
In which cases we'll get `couldn't connect to host` error? I suppose only, when selenium is misconfigured, which is an error as well.

gggeek replied:
1. no idea. I left it as I found it
2. agree. I just did not want to change too much stuff

gggeek replied:
point 2 done

gggeek commented:
ps: will squash after all the changes from suggestions

aik099 commented:
> ps: will squash after all the changes from suggestions

Since GitHub code review process isn't very feature reach, then please reply to inline comments once you'll have corresponding changes committed/pushed.

gggeek commented:
All points for which there was a clear indication have been tackled.
A couple of points remain open for which I have no clear answer on best solution.
Waiting for Travis to validate the changes (https://travis-ci.org/instaclick/php-webdriver/builds/67275002).

aik099 commented:
While this PR is being reviewed the 0c20707dcf30a32728fd6bdeeab996c887fdb2fb commit was made to revert code line added, that caused exception.

robocoder commented on lib/WebDriver/Exception/CurlDetails.php:
```diff
@@ -0,0 +1,23 @@
+namespace WebDriver\Exception;
+
+
+final class CurlDetails extends \Exception {
```
Move to lib/WebDriver/Service/CurlServiceException.php and rename class to CurlServiceException

robocoder commented on lib/WebDriver/Service/CurlService.php:
```diff
@@ -24,6 +24,7 @@
 namespace WebDriver\Service;
 
 use WebDriver\Exception as WebDriverException;
+use WebDriver\Exception\CurlDetails as CurlDetailsException;
```
Remove the two `use` statements.

robocoder commented on lib/WebDriver/Service/CurlService.php:
```diff
@@ -90,17 +91,25 @@ public function execute($requestMethod, $url, $parameters = null, $extraOptions
-            throw WebDriverException::factory(WebDriverException::CURL_EXEC, $message . "\n\n" . $error);
+            $detailsException = new CurlDetailsException($error, $errno, null, $rawResult, $info);
+            throw WebDriverException::factory(WebDriverException::CURL_EXEC, $message, $detailsException);
         }
```
Dropping the WebDriverException dependency is a good idea. Let's just `throw new CurlServiceException(...)`

robocoder commented on lib/WebDriver/AbstractWebDriver.php:
```diff
@@ -122,15 +124,28 @@ protected function curl($requestMethod, $command, $parameters = null, $extraOpti
+        // According to https://w3c.github.io/webdriver/webdriver-spec.html all 4xx responses are to be considered an error and return plaintext,
+        // while 5xx responses are json encoded
+        if (substr($info['http_code'], 0, 1) === '4') {
+            throw WebDriverException::factory(Exception::CURL_EXEC, 'Webdriver http error: ' . $info['http_code'] . ', payload :' . substr($rawResult, 0, 1000));
```
Wrong namespace. Change the occurrences of Exception::CURL_EXEC to WebDriverException::CURL_EXEC

gggeek commented:
Implemented all changes suggested by @robocoder and squashed.
I suggest that the improvements to Travis (testing against multiple versions of Selenium/FF) be carried out in a separate issue/PR.

chillu commented:
@robocoder Is there anything blocking the merge of this? We're using php-webdriver with Behat (which requires `~1.1` of your library), and the 1.4.3 release made it much harder to debug any failures in CI due to important context being swallowed as described in #64. Also, once you merge this, can you please tag a new 1.4.4 release as well?

robocoder commented:
@chillu Sorry. The last commit message was "WIP: ...", and it simply fell by the wayside. If everyone is satisfied with this PR's functionality, I'll review and merge.

chillu commented:
@robocoder I think you can take that silence as agreement? ;)

gggeek commented:
I am also ok with this being merged :-)

chillu commented:
@robocoder Given your message from October, are you happy to press the big green button? ;)

tractorcow commented:
@robocoder this merge caused breaking regressions. Can you please look at https://github.com/instaclick/php-webdriver/pull/79?
